src/lib.rs

```diff
@@ -279,22 +287,41 @@ pub mod pallet {
                 log::error!("Header error. Signature or token id is not correct!");
             }
             ensure!(res, "header error. Please check the logs.");
-
-            // checked add bug in example cod where it uses checked_sub
-            total_output = total_output
-                .checked_add(output.value)
-                .ok_or("output value overflow")?;
             new_utxos.push(hash.as_fixed_bytes().to_vec());
         }
 
         // if no race condition, check the math
         if missing_utxos.is_empty() {
-            ensure!(
-                total_input >= total_output,
-                "output value must not exceed input value"
-            );
-            reward = total_input
-                .checked_sub(total_output)
+            // We have to check sum of input tokens is less or equal to output tokens.
+            let mut inputs_sum: BTreeMap<TokenID, Value> = BTreeMap::new();
+            let mut outputs_sum: BTreeMap<TokenID, Value> = BTreeMap::new();
+
+            for x in input_vec {
+                let value =
+                    x.1.checked_add(*inputs_sum.get(&x.0).unwrap_or(&0))
+                        .ok_or("input value overflow")?;
+                inputs_sum.insert(x.0, value);
+            }
+            for x in out_vec {
+                let value =
+                    x.1.checked_add(*outputs_sum.get(&x.0).unwrap_or(&0))
+                        .ok_or("output value overflow")?;
+                outputs_sum.insert(x.0, value);
+            }
+
+            for output_token in &outputs_sum {
+                match inputs_sum.get(&output_token.0) {
+                    Some(input_value) => ensure!(
+                        input_value >= &output_token.1,
+                        "output value must not exceed input value"
+                    ),
+                    None => frame_support::fail!("input for the token not found"),
+                }
+            }
+
+            // Reward at the moment only in MLT
+            reward = inputs_sum[&(crate::TokenType::MLT as TokenID)]
```

Comment:
Looks good to me for handling MLT rewards in the testnet since we will enforce the MLT tx fee.
My only concern is the scaling of the crate::TokenType as we'd need to maintain a massive structure of all tokens for this to scale but that's not an issue for this pr. We can deal with that later. 

Looks good to me as is

@BCarlaYap look good to you too?

Reply:
I think we need to make functionality to create tokens in the UI and after that, perhaps in the storage, we can save Vec with tokens info. And of course, won't use any concrete enums for it due to list of tokens must be flexible.

Reply:
Does this mean a non-MLT fee is currently destroyed?

Reply:
Short answer - yes! Have a look [here please ](https://github.com/mintlayer/pallet-utxo/pull/14#discussion_r694926008).

Reply:
Yep cool no worries for now, just wanted to make sure I understood it